jetty-deploy/src/main/java/org/eclipse/jetty/deploy/providers/ScanningAppProvider.java

```diff
@@ -152,12 +155,49 @@ protected void doStart() throws Exception
         _scanner.setReportDirs(true);
         _scanner.setScanDepth(1); //consider direct dir children of monitored dir
         _scanner.addListener(_scannerListener);
+        _scanner.setReportExistingFilesOnStartup(true);
 
-        addBean(_scanner);
+        if (_deployOnStartup)
+        {
+            addBean(_scanner);
+        }
+        else
+        {
+            // Setup listener to wait for Server in STARTED state, which
+            // triggers the first scan of the monitored directories
+            getDeploymentManager().getServer().addEventListener(getListener());
+        }
 
         super.doStart();
     }
 
+    private LifeCycle.Listener getListener()
+    {
+        ContainerLifeCycle container = this;
+        return new LifeCycle.Listener()
+        {
+            @Override
+            public void lifeCycleStarted(LifeCycle event)
+            {
+                if (event instanceof Server)
+                {
+                    if (LOG.isDebugEnabled())
+                        LOG.debug("Triggering Delayed Scan of {}", _monitored);
+                    try
+                    {
+                        _scanner.start();
+                    }
+                    catch (Exception e)
+                    {
+                        throw new RuntimeException("Unable to start Scanner", e);
+                    }
+                    container.addBean(_scanner);
```

Comment:
You want to add the scanner as a managed bean so that it will be stopped with the provider.  I think it is safe and better to add it before starting it.

```suggestion
                    container.addBean(_scanner, true);
                    try
                    {
                        _scanner.start();
                    }
                    catch (Exception e)
                    {
                        throw new RuntimeException("Unable to start Scanner", e);
                    }
```

Reply:
No longer relevant.